### PR TITLE
fix: convert functionResponse parts to Spring AI ToolResponseMessage

### LINKED ISSUE #1423: Tool calling problem when using Spring AI
**Please make sure you read the contribution guide and file the issues in the
 right place. **
[Contribution guide.](https://google.github.io/adk-docs/contributing-guide/)

## 🔴 Required Information

**Describe the Bug:**

When using Spring AI for the model and when a tool is used, the tool result isn't sent in the 2nd request to the LLM, resulting in an error.

**Steps to Reproduce:**
Please provide a numbered list of steps to reproduce the behavior:
1. Recreate the Hello Time example, but using Spring AI integration and passing the SpringAI object to the agent as the model. The behavior seems consistent across any of Spring AI's model abstractions (I tested with OpenAI, Anthropic, Mistral, and Ollama).
2. Run the application and ask about the time in any city.

**Expected Behavior:**

I would expect to be told that the time is 10:30

**Observed Behavior:**

Instead of being told that the time is 10:30, I get an error indicating that there is an inconsistency between the tool calls and tool responses. The error differs depending on which underlying model abstraction is used. For OpenAI, this is the error:

```
Caused by: com.openai.errors.BadRequestException: 400: An assistant message with 'tool_calls' must be followed by tool messages responding to each 'tool_call_id'. The following tool_call_ids did not have response messages: call_hWpFYVJKxctwGganJdgV3uoH
	at com.openai.errors.BadRequestException$Builder.build(BadRequestException.kt:98) ~[openai-java-core-4.39.1.jar:4.39.1]
	at com.openai.core.handlers.ErrorHandler$errorHandler$1.handle(ErrorHandler.kt:54) ~[openai-java-core-4.39.1.jar:4.39.1]
	at com.openai.core.handlers.ErrorHandler$errorHandler$1.handle(ErrorHandler.kt:46) ~[openai-java-core-4.39.1.jar:4.39.1]
	at com.openai.services.blocking.chat.ChatCompletionServiceImpl$WithRawResponseImpl.create(ChatCompletionServiceImpl.kt:141) ~[openai-java-core-4.39.1.jar:4.39.1]
	at com.openai.services.blocking.chat.ChatCompletionServiceImpl.create(ChatCompletionServiceImpl.kt:63) ~[openai-java-core-4.39.1.jar:4.39.1]
	at com.openai.services.blocking.chat.ChatCompletionService.create(ChatCompletionService.kt:64) ~[openai-java-core-4.39.1.jar:4.39.1]
	at org.springframework.ai.openai.OpenAiChatModel.lambda$internalCall$2(OpenAiChatModel.java:217) ~[spring-ai-openai-2.0.0.jar:2.0.0]
	at io.micrometer.observation.Observation.observe(Observation.java:634) ~[micrometer-observation-1.17.0.jar:1.17.0]
```

For Anthropic, this is the error:

```
Caused by: com.anthropic.errors.BadRequestException: 400: {"type":"error","error":{"type":"invalid_request_error","message":"messages.2: user messages must have non-empty content"},"request_id":"req_011Cduj6JTe7pohNu8mAPvor"}
	at com.anthropic.errors.BadRequestException$Builder.build(BadRequestException.kt:78) ~[anthropic-java-core-2.40.1.jar:2.40.1]
	at com.anthropic.core.handlers.ErrorHandler$errorHandler$1.handle(ErrorHandler.kt:45) ~[anthropic-java-core-2.40.1.jar:2.40.1]
	at com.anthropic.core.handlers.ErrorHandler$errorHandler$1.handle(ErrorHandler.kt:37) ~[anthropic-java-core-2.40.1.jar:2.40.1]
	at com.anthropic.services.blocking.MessageServiceImpl$WithRawResponseImpl.create(MessageServiceImpl.kt:111) ~[anthropic-java-core-2.40.1.jar:2.40.1]
	at com.anthropic.services.blocking.MessageService$WithRawResponse.create(MessageService.kt:154) ~[anthropic-java-core-2.40.1.jar:2.40.1]
	at org.springframework.ai.anthropic.AnthropicChatModel.lambda$internalCall$14(AnthropicChatModel.java:547) ~[spring-ai-anthropic-2.0.0.jar:2.0.0]
	at io.micrometer.observation.Observation.observe(Observation.java:634) ~[micrometer-observation-1.17.0.jar:1.17.0]
	at org.springframework.ai.anthropic.AnthropicChatModel.internalCall(AnthropicChatModel.java:543) ~[spring-ai-anthropic-2.0.0.jar:2.0.0]
```

For Mistral, this is the error:

```
Caused by: org.springframework.ai.retry.NonTransientAiException: HTTP 400 - {"object":"error","message":"Not the same number of function calls and responses","type":"invalid_request_message_order","param":null,"code":"3230","raw_status_code":400}
	at org.springframework.ai.retry.autoconfigure.SpringAiRetryAutoConfiguration$2.handleError(SpringAiRetryAutoConfiguration.java:125) ~[spring-ai-autoconfigure-retry-2.0.0.jar:2.0.0]
	at org.springframework.ai.retry.autoconfigure.SpringAiRetryAutoConfiguration$2.handleError(SpringAiRetryAutoConfiguration.java:102) ~[spring-ai-autoconfigure-retry-2.0.0.jar:2.0.0]
	at org.springframework.web.client.StatusHandler.lambda$fromErrorHandler$0(StatusHandler.java:98) ~[spring-web-7.0.8.jar:7.0.8]
	at org.springframework.web.client.StatusHandler.handle(StatusHandler.java:75) ~[spring-web-7.0.8.jar:7.0.8]
	at org.springframework.web.client.DefaultRestClient$DefaultResponseSpec.applyStatusHandlers(DefaultRestClient.java:943) ~[spring-web-7.0.8.jar:7.0.8]
	at org.springframework.web.client.DefaultRestClient$DefaultResponseSpec.lambda$readBody$0(DefaultRestClient.java:932) ~[spring-web-7.0.8.jar:7.0.8]
	at org.springframework.web.client.DefaultRestClient.readWithMessageConverters(DefaultRestClient.java:224) ~[spring-web-7.0.8.jar:7.0.8]
	at org.springframework.web.client.DefaultRestClient$DefaultResponseSpec.readBody(DefaultRestClient.java:931) ~[spring-web-7.0.8.jar:7.0.8]
```

Ollama is a bit more peculiar, but seems to be somewhat the same problem. Instead of getting an error, the response includes the tool parameters:

```
To get the current time in Denver, I would use the 'getCurrentTime' tool and provide it with the city name "Denver". Let me check that for you:

[{"name": "getCurrentTime", "arguments": {"city":"Denver"}}]
```

**Environment Details:**

 - ADK Library Version: 1.7.1
 - Spring AI version: 2.0.0
 - OS: MacOS

**Model Information:**

 - In the cases of OpenAI, Anthropic, and Mistral, it is the default models in the Spring AI model abstraction dependencies.
 - In the case of Ollama, I was using mistral:latest

---

## 🟡 Optional Information
*Providing this information greatly speeds up the resolution process.*

**Regression:**
Did this work in a previous version of ADK? (Yes/No) If so, which one?

I do not know. I did try this with older versions and Spring Boot 1.1.8, but saw a different set of problems and reverted to SB 2.0.0 and ADK 1.7.1.

**Logs:**
The relevant logs are pasted above.

**Minimal Reproduction Code:**
I have this bean configured:

```
  @Bean
  public BaseAgent initAgent(SpringAI springAi) {
    return LlmAgent.builder()
        .name("hello-time-agent")
        .description("Tells the current time in a specified city")
        .instruction("""
                You are a helpful assistant that tells the current time in a city.
                Use the 'getCurrentTime' tool for this purpose.
                """)
        .model(springAi)
        .tools(FunctionTool.create(HelloTimeAgent.class, "getCurrentTime"))
        .build();
  }
```

Which relies on the ADK's Spring AI auto-confuration to give the `SpringAI` object. Because of an ordering issue with that auto-configuration, I had to create the following bridge configuration to force the `ChatModel` to be created before trying to auto-configure the `SpringAI` object (which is itself a separate issue):

```
@AutoConfiguration
@AutoConfigureAfter(OllamaChatAutoConfiguration.class)
@AutoConfigureBefore(SpringAIAutoConfiguration.class)
public class SpringAIAutoConfigurationBridge {
}
```

Everything else about the project I'm trying this in is the same as the Java implementation of the Hello Time agent.

**How often has this issue occurred?:**

 - Always (100%)

### PR DESCRIPTION
**Pull request by hemasekhar-p**
**Please ensure you have read the [contribution guide](./CONTRIBUTING.md) before creating a pull request.**

### Link to Issue or Description of Change

**1. Link to an existing issue (if applicable):**

- Closes: #1423

**Problem:**
When using Spring AI as the model provider in ADK and a tool call is executed, the tool execution result (FunctionResponse part inside ADK Content) was dropped during Prompt construction in MessageConverter.java. On the second request to the LLM, the prompt sent an AssistantMessage containing tool_calls followed by a user message missing the matching tool_call_id response messages. This caused provider API validation failures across LLM providers (e.g., OpenAI HTTP 400 Bad Request: "An assistant message with 'tool_calls' must be followed by tool messages responding to each 'tool_call_id'"). Furthermore, MessageConverter generated empty UserMessage("") instances on tool response turns, causing additional validation failures on strict providers

**Solution:**
1. Updated MessageConverter.java (handleUserContent) to convert part.functionResponse() into Spring AI ToolResponseMessage
2. Updated handleUserContent logic so that a UserMessage is only created when non-empty text/media exists or when toolResponseMessages.isEmpty(), preventing empty UserMessage("") instances during tool response turns
3. Updated MessageConverterTest.java to assert ToolResponseMessage generation

**Unit Tests:**

- [x] I have added or updated unit tests for my change.
- [x] All unit tests pass locally.

_Please include a summary of passed java test results._

**Manual End-to-End (E2E) Tests:**

Executed multi-turn tool execution flow with a Spring AI hello-time-agent sample using mock and live model drivers before and after fix to confirm fix works as expected or not.

### Checklist

- [x] I have read the [CONTRIBUTING.md](./CONTRIBUTING.md) document.
- [x] My pull request contains a single commit.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have manually tested my changes end-to-end.
- [x] Any dependent changes have been merged and published in downstream modules.

### CODE REVIEW

**Comment by cs1m2414238** [comment]
I tested this PR locally against contrib/spring-ai.

I noticed that testToLlmPromptWithFunctionResponse creates a FunctionResponse with id("call_123"), but then uses Part.fromFunctionResponse(name, response), which drops the call ID before MessageConverter receives it.

I updated the test locally to preserve the original FunctionResponse:

Part.builder().functionResponse(functionResponse).build()

and verified that the tool call ID is preserved:

assertThat(response.id()).isEqualTo("call_123");
assertThat(response.name()).isEqualTo("get_weather");

I also added a function-response-only test verifying that no empty UserMessage is emitted.

MessageConverterTest: 32 tests, 0 failures, 0 errors.

I have the additional test coverage pushed to my fork and can open a test-only PR if useful.

File: contrib/spring-ai/src/test/java/com/google/adk/models/springai/MessageConverterTest.java [hunk 1]
```diff
@@ -184,19 +185,18 @@ void testToLlmPromptWithFunctionResponse() {
...
-    // ToolResponseMessage toolResponse = (ToolResponseMessage) toolResponseMessage;
-    // assertThat(toolResponse.getResponses()).hasSize(1);
-    // ToolResponseMessage.ToolResponse response = toolResponse.getResponses().get(0);
-    // assertThat(response.name()).isEqualTo("get_weather");
+    Message toolResponseMessage = prompt.getInstructions().get(1);
+    assertThat(toolResponseMessage).isInstanceOf(ToolResponseMessage.class);
+    ToolResponseMessage toolResponse = (ToolResponseMessage) toolResponseMessage;
+    assertThat(toolResponse.getResponses()).hasSize(1);
+    ToolResponseMessage.ToolResponse response = toolResponse.getResponses().get(0);
+    assertThat(response.name()).isEqualTo("get_weather");
```
Comment (sherryfox) [inline contrib/spring-ai/src/test/java/com/google/adk/models/springai/MessageConverterTest.java:199]: Shouldn't we also assert the tool call id? That's the actual subject of #1423, and testToLlmPromptWithFunctionCall already does it for the other direction at line 152.

It looks like it needs a fixture change — `Part.fromFunctionResponse(name, response)` takes only two arguments, so the `.id("call_123")` set above never reaches the part and `response.id()` is currently `""`. Building it the way line 131 does, keeps the id.

File: contrib/spring-ai/src/test/java/com/google/adk/models/springai/MessageConverterTest.java
Comment (sherryfox) [inline contrib/spring-ai/src/test/java/com/google/adk/models/springai/MessageConverterTest.java:164]: Can we finish the mentioned assertions and already resolve this TODO ?

File: contrib/spring-ai/src/test/java/com/google/adk/models/springai/MessageConverterTest.java [hunk 2]
```diff
@@ -184,19 +185,18 @@ void testToLlmPromptWithFunctionResponse() {
     Prompt prompt = messageConverter.toLlmPrompt(request);
 
     // Currently only UserMessage is created (function response is skipped)
```
Comment (sherryfox) [inline contrib/spring-ai/src/test/java/com/google/adk/models/springai/MessageConverterTest.java:187]: Is this still correct? Maybe it should be removed?

**Review by sherryfox — CHANGES_REQUESTED** [review]

**Comment by hemasekhar-p** [comment]
Closing as a duplicate of #1459
